SqlServer/Import-BacpacToSqlServer/Import-BacpacToSqlServer.psm1

```diff
@@ -160,35 +215,53 @@ function Import-BacpacToSqlServer
 
 
 
-        # And now we get to actually importing the database after constructing a connection string with validated details.
-        $ConnectionString = "Data Source=$DataSource;Initial Catalog=$DatabaseName;Integrated Security=True;"
-        & "$sqlPackageExecutablePath" /Action:Import /SourceFile:"$BacpacPath" /TargetConnectionString:"$ConnectionString"
+        [Collections.ArrayList] $parameters = @(
+            '/Action:Import',
+            "/TargetServerName:`"$SqlServerName`"",
+            "/TargetDatabaseName:`"$DatabaseName`"",
+            '/TargetTrustServerCertificate:True',
+            '/TargetTimeout:0',
+            "/SourceFile:`"$BacpacPath`""
+        )
+
+        if (![string]::IsNullOrEmpty($Username) -and ![string]::IsNullOrEmpty($Password))
+        {
+            $parameters += "/TargetUser:`"$Username`""
+            $parameters += "/TargetPassword:`"$Password`""
+        }
+
+        # And now we get to actually importing the database after setting up all the necessary parameters.
+        & "$sqlPackageExecutablePath" @parameters
 
 
 
         # Importing the database was successful.
         if ($LASTEXITCODE -eq 0)
         {
-            # If there was a database with an identical name on the target server, we'll swap it with the one we've just imported.
+            # If there was a database with an identical name on the target server, we'll swap it with the one we've just
+            # imported.
             if ($databaseExists)
             {
-                $server = New-Object ('Microsoft.SqlServer.Management.Smo.Server') $DataSource
+                $server = New-Object ('Microsoft.SqlServer.Management.Smo.Server') (New-SqlServerConnection $SqlServerName $UserName $securePassword)
                 $server.KillAllProcesses($originalDatabaseName)
                 $server.Databases[$originalDatabaseName].Drop()
```

Comment:
Now I get this exception:
```
Enabling index 'IDX_TimeFieldIndex_ContentType'.
Enabling index 'IDX_TimeFieldIndex_Latest'.
Enabling index 'IDX_TimeFieldIndex_Published'.
Enabling index 'IDX_TimeFieldIndex_Time'.
Enabling index 'IDX_UserIndex_IsEnabled'.
Successfully imported database.
Time elapsed 0:00:24.38
Exception calling "Drop" with "0" argument(s): "Drop failed for Database 'master'. "
Press Enter to continue...
```

I've printed out the value of `$originalDatabaseName` and it's `master`. It makes sense that you can't drop the `master`, since it's a system database.

This value is coming from the connection string passed in from the command line arguments. I think it sticks because of [this change](https://github.com/Lombiq/Utility-Scripts/pull/45/files#diff-e7e059c1c0580744969af78076e6c91fca42b2a639d7a38a6798b8777b6726cbR168-R172). The database name is only updated from the file name conditionally if it's not already specified. But the `Database` segment is always in the connection string, so this condition will never be true.,

Reply:
That change was there for a long time, but the connection string in `Import-BacpacToSqlServer.docker.edit.reg` has the `Database=master` segment in it, which wasn't taken into account before (because we only checked `Initial Catalog`, not `Database`).
I tested with Docker before, just not this registry script specifically, but it's an easy fix, so I've removed that segment, so it uses the file name.